### PR TITLE
ARROW-4503: [C#] Eliminate allocations in ArrowStreamReader when reading from a Stream

### PR DESCRIPTION
**Pull request by eerhardt**
When reading `RecordBatch` instances from a .NET `Stream` using the `ArrowStreamReader` class, it is currently allocating and copying memory 3 times for the data.

I've changed it to take a MemoryPool in the constructor, `ArrowStreamReader` will allocate from the pool a single time for each `RecordBatch` it reads, copy from the `Stream` into that memory once, and then use it in the returned `RecordBatch`.

I've also fixed the comments from #3736 that were deemed to be better fixed with this change.

### CODE REVIEW

**Comment by eerhardt** [comment]
Here are some preliminary perf results of this change:

|                      Method |      Mean |     Error |    Gen 0/1k Op | Allocated Memory/Op |
|---------------------------- |----------:|----------:|----------:|--------------------:|
| **BEFORE** ArrowReaderWithMemoryStream | 81.417 ms | 1.6083 ms | 3000.0000 |         84000.73 KB |
| **AFTER** ArrowReaderWithMemoryStream | 29.431 ms | 0.5826 ms |             - |             6.28 KB |

The reason `Allocated Memory/Op` drops to almost 0 is because we are allocating from the default `MemoryPool`, which pulls from native memory, which isn't tracked by BenchmarkDotNet.

**Comment by chutchinson** [comment]
@eerhardt Is there a way to report the native memory to the benchmark? Seems like it would be valuable for benchmarks, especially considering the current allocator is naive and probably even broken and/or not thread-safe. I expect it will be replaced or supplemented with Jemalloc.NET or some other more reliable allocation mechanism.

**Comment by eerhardt** [comment]
> Is there a way to report the native memory to the benchmark?

I'm not sure, but maybe @adamsitnik might be able to answer.

On a related topic, @chutchinson, is there a specific reason why the default MemoryPool uses native memory? I guess my assumption was that the default should just be regular `byte[]` instances, which could be garbage collected. I also noticed that there doesn't seem to be a way to return memory back to the pool to be reused. Is that intentional?

**Comment by eerhardt** [comment]
@stephentoub @chutchinson @pgovind - this is now ready for review.

File: csharp/src/Apache.Arrow/Extensions/StreamExtensions.netcoreapp2.1.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+using System;
+using System.IO;
+
+namespace Apache.Arrow
+{
+    // Helpers to read from Stream to Memory<byte> on netcoreapp
+    internal static class StreamExtensions
+    {
+        public static int Read(this Stream stream, Memory<byte> buffer)
```
Comment (stephentoub) [inline csharp/src/Apache.Arrow/Extensions/StreamExtensions.netcoreapp2.1.cs]: Why is this necessary?  Why not just call it with buffer.Span at the call site?

Comment (stephentoub) [reply csharp/src/Apache.Arrow/Extensions/StreamExtensions.netcoreapp2.1.cs]: I see. You've added this as an abstraction to enable fishing out the array when the span-based overload doesn't exist on netstandard. Ok.

File: csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs [hunk 2]
```diff
@@ -70,36 +72,79 @@ protected override async Task ReadSchemaAsync()
             {
                 BaseStream.Position = BaseStream.Length - ArrowFileConstants.Magic.Length - 4;
 
-                bytesRead = await BaseStream.ReadAsync(buffer, 0, 4).ConfigureAwait(false);
-                footerLength = BinaryPrimitives.ReadInt32LittleEndian(buffer);
+                bytesRead = await BaseStream.ReadAsync(buffer.AsMemory(0, 4)).ConfigureAwait(false);
```
Comment (stephentoub) [inline csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs]: You'd be better off moving the declaration of bytesRead into each of these delegates so they don't have to close over the shared field and instead it can just be a local (that doesn't need to be lifted to any state machine).

File: csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs [hunk 3]
```diff
@@ -70,36 +72,79 @@ protected override async Task ReadSchemaAsync()
...
+                footerLength = ReadFooterLength(buffer, bytesRead);
+            }).ConfigureAwait(false);
 
-                if (bytesRead != 4) throw new InvalidDataException(
-                    $"Failed to read footer length. Read <{bytesRead}>, expected 4.");
+            await Buffers.RentReturnAsync(footerLength, async (buffer) =>
+            {
+                _footerStartPostion = (int)BaseStream.Length - footerLength - ArrowFileConstants.Magic.Length - 4;
```
Comment (stephentoub) [inline csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs]: Is buffer.Length == footerLength?  If so, you could use that instead of footerLength to avoid closing over footerLength.  If instead buffer is a byte[] rented from the pool and thus could have a length > footerLength, you might change the signature to pass in a `Memory<byte>` rather than a `byte[]`, with it sliced to the right length, and then you could use `buffer.Length` for that purpose.

File: csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs [hunk 4]
```diff
@@ -26,12 +27,14 @@ internal class ArrowStreamReaderImplementation : ArrowReaderImplementation
...
+        private readonly MemoryPool _memoryPool;
 
-        public ArrowStreamReaderImplementation(Stream stream, bool leaveOpen)
+        public ArrowStreamReaderImplementation(Stream stream, MemoryPool memoryPool, bool leaveOpen)
         {
             BaseStream = stream;
+            _memoryPool = memoryPool ?? MemoryPool.Default.Value;
             _leaveOpen = leaveOpen;
-            Buffers = ArrayPool<byte>.Create();
+            Buffers = ArrayPool<byte>.Shared;
```
Comment (stephentoub) [inline csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs]: Does it need to be a property?  It'd be more efficient to just have all callers use `ArrayPool<byte>.Shared.Rent/Return`, both because it would remove the extra field from this instance and because the .NET Core JIT will likely be able to devirtualize those calls.

File: csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs [hunk 5]
```diff
@@ -51,69 +54,101 @@ public override async Task<RecordBatch> ReadNextRecordBatchAsync(CancellationTok
...
+            RecordBatch result = null;
+            await Buffers.RentReturnAsync(messageLength, async (messageBuff) =>
+            {
+                bytesRead += await BaseStream.ReadAsync(messageBuff.AsMemory(0, messageLength), cancellationToken)
+                    .ConfigureAwait(false);
+                Flatbuf.Message message = Flatbuf.Message.GetRootAsMessage(new FlatBuffers.ByteBuffer(messageBuff));
 
-                lengthBuffer = Buffers.Rent(4);
-                bytesRead += await BaseStream.ReadAsync(lengthBuffer, 0, 4, cancellationToken)
+                Memory<byte> bodyBuff = _memoryPool.Allocate((int)message.BodyLength);
```
Comment (stephentoub) [inline csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs]: Where and when do these buffers get returned to the pool?

**Comment by stephentoub** [comment]
> ArrowReaderWithMemoryStream

Do we expect this to be a common case? If so, have you considered extracting the byte[] underlying the MemoryStream and using the memory-based reader instead of stream-based? Your need to read the header info at least presumably to know how to slice it, but that should then should then allow that case to be zero-copy, too.

**Comment by chutchinson** [comment]
> On a related topic, @chutchinson, is there a specific reason why the default MemoryPool uses native memory? I guess my assumption was that the default should just be regular `byte[]` instances, which could be garbage collected. I also noticed that there doesn't seem to be a way to return memory back to the pool to be reused. Is that intentional?

The default memory pool allocates native memory because the allocated pointer can be adjusted to satisfy the Arrow memory alignment rule in a cross-platform way without GC pinning and the associated complexity. Admittedly I don't have a sufficiently deep understanding of the CLR (especially CoreCLR) to imagine a better way. 

There needs to be more thought and care put into the design of the memory allocation API and implementation. There is no mechanism for returning allocated memory; at the moment it is garbage collected as appropriate via the MemoryManager<T> API.

**Comment by stephentoub** [comment]
> There is no mechanism for returning allocated memory; at the moment it is garbage collected as appropriate via the MemoryManager API.

Given this, I'd recommend not using this pool in this PR.  It's going to make things significantly less efficient.

File: csharp/src/Apache.Arrow/Extensions/StreamExtensions.netcoreapp2.1.cs (hunk 1 above)
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Extensions/StreamExtensions.netcoreapp2.1.cs]: Is there are better way to do this? I didn't find one.

File: csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs (hunk 2 above)
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs]: Good call. Fixed.

File: csharp/src/Apache.Arrow/Extensions/StreamExtensions.netcoreapp2.1.cs (hunk 1 above)
Comment (stephentoub) [reply csharp/src/Apache.Arrow/Extensions/StreamExtensions.netcoreapp2.1.cs]: Given the desire to also compile for netstandard, this seems reasonable.

File: csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs [hunk 6]
```diff
@@ -70,36 +72,79 @@ protected override async Task ReadSchemaAsync()
...
+            }
+
+            ValidateFile();
+
+            var bytesRead = 0;
+            var footerLength = 0;
+
+            Buffers.RentReturn(4, (buffer) =>
+            {
+                BaseStream.Position = BaseStream.Length - ArrowFileConstants.Magic.Length - 4;
```
Comment (bkietz) [inline csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs]: `BaseStream.Length - ArrowFileConstants.Magic.Length - 4` appears several times, could it be extracted into a well named method?

**Comment by adamsitnik** [comment]
> Is there a way to report the native memory to the benchmark?

File: csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs (hunk 6 above)
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs]: This was fixed in #4017.

File: csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs (hunk 5 above)
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs]: I have changed this to return an `IMemoryOwner<byte>` which is then wrapped by `IDisposable` objects. When those objects (`RecordBatch`, `ArrowBuffer`) get disposed, they dispose the `IMemoryOwner`.

File: csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs (hunk 4 above)
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs]: Fixed.

**Comment by eerhardt** [comment]
OK - I finally got back to this PR and it is ready for re-review. I believe I have addressed all feedback and concerns. Please let me know.

Here are some updated numbers with the new benchmarks added:

## Before
|                      Method |   Count |         Mean |        Error |      StdDev | Gen 0/1k Op | Allocated Memory/Op |
|---------------------------- |-------- |-------------:|-------------:|------------:|------------:|--------------------:|
| ArrowReaderWithMemoryStream |   10000 |    639.69 us |    25.797 us |    71.91 us |   3000.0000 |           947.22 KB |
|       ArrowReaderWithMemory |   10000 |     59.98 us |     6.349 us |    17.70 us |           - |             4.48 KB |
| ArrowReaderWithMemoryStream | 1000000 | 78,008.70 us | 1,536.342 us | 1,437.10 us |   3000.0000 |         83999.47 KB |
|       ArrowReaderWithMemory | 1000000 |  2,675.14 us |   119.735 us |   333.77 us |           - |             4.48 KB |

## After
|                                    Method |   Count |         Mean |      Error |    StdDev | Gen 0/1k Op |  Allocated Memory/Op |
|------------------------------------------ |-------- |-------------:|-----------:|----------:|------------:|--------------------:|
|               ArrowReaderWithMemoryStream |   10000 |    185.59 us |  22.262 us |  65.29 us |   1000.0000 |              5.2 KB |
| ArrowReaderWithMemoryStream_ManagedMemory |   10000 |    110.95 us |  11.195 us |  32.12 us |           - |             5.19 KB |
|                     ArrowReaderWithMemory |   10000 |     55.50 us |   5.738 us |  16.37 us |           - |              4.7 KB |
|               ArrowReaderWithMemoryStream | 1000000 | 34,292.18 us | 582.247 us | 544.63 us |   1000.0000 |              5.2 KB |
| ArrowReaderWithMemoryStream_ManagedMemory | 1000000 | 27,583.96 us | 600.099 us | 616.26 us |           - |          41997.4 KB |
|                     ArrowReaderWithMemory | 1000000 |  2,897.50 us | 157.003 us | 462.93 us |           - |              4.7 KB |

Notes on the benchmarks:
1. I couldn't do the `_ManagedMemory` tests before this change, since that API didn't exist.
2. Notice the gains on `ArrowReaderWithMemoryStream_ManagedMemory` with `Count=10,000`. It is a 40% gain vs. the default case. The reasoning is because the `_ManagedMemory` uses `MemoryPool<byte>` to allocate managed memory. In the case of `10,000` records (the default in Spark), we are able to pool the memory buffers and don't need to allocate a new buffer each iteration.

File: csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs [hunk 7]
```diff
@@ -78,23 +79,24 @@ protected async ValueTask<RecordBatch> ReadRecordBatchAsync(CancellationToken ca
...
                 Flatbuf.Message message = Flatbuf.Message.GetRootAsMessage(CreateByteBuffer(messageBuff));
 
-                await Buffers.RentReturnAsync((int)message.BodyLength, async (bodyBuff) =>
-                {
-                    int bodyBytesRead = await BaseStream.ReadFullBufferAsync(bodyBuff, cancellationToken)
-                        .ConfigureAwait(false);
-                    EnsureFullRead(bodyBuff, bodyBytesRead);
+                int bodyLength = checked((int)message.BodyLength);
+
+                IMemoryOwner<byte> bodyBuffOwner = _allocator.Allocate(bodyLength);
```
Comment (adamsitnik) [inline csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs]: again: is it acceptable to have `bodyBuffOwner` be a null?

what is the outcome of reading into an empty memory?

File: csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs [hunk 8]
```diff
@@ -78,23 +79,24 @@ protected async ValueTask<RecordBatch> ReadRecordBatchAsync(CancellationToken ca
...
+                Memory<byte> bodyBuff = bodyBuffOwner?.Memory.Slice(0, bodyLength) ?? Memory<byte>.Empty;
+                bytesRead = await BaseStream.ReadFullBufferAsync(bodyBuff, cancellationToken)
+                    .ConfigureAwait(false);
+                EnsureFullRead(bodyBuff, bytesRead);
 
-                    FlatBuffers.ByteBuffer bodybb = CreateByteBuffer(bodyBuff);
-                    result = CreateArrowObjectFromMessage(message, bodybb);
-                }).ConfigureAwait(false);
+                FlatBuffers.ByteBuffer bodybb = CreateByteBuffer(bodyBuff);
+                result = CreateArrowObjectFromMessage(message, bodybb, bodyBuffOwner);
```
Comment (adamsitnik) [inline csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs]: should the owner be disposed here?

edit: I can see that `RecordBatch` is disposable now. Is the Arrow deserialization lazy and we need to keep the memory alive until somebody reads from it?

File: csharp/src/Apache.Arrow/RecordBatch.cs [hunk 9]
```diff
@@ -40,6 +42,24 @@ public IArrowArray Column(string columnName)
             return _arrays[fieldIndex];
         }
 
+        public void Dispose()
```
Comment (adamsitnik) [inline csharp/src/Apache.Arrow/RecordBatch.cs]: are there any types that derive from `RecordBatch`? Do we need full dispose pattern implementation here?

File: csharp/test/Apache.Arrow.Tests/ArrowFileReaderTests.cs [hunk 10]
```diff
@@ -47,6 +48,55 @@ public void Ctor_LeaveOpenTrue_StreamValidOnDispose()
...
 
+        [Fact]
+        public async Task Ctor_MemoryPool_AllocatesFromPool()
+        {
+            await Ctor_MemoryPool_AllocatesFromPoolHelper(
+                shouldLeaveOpen: false,
+                (stream, memoryPool) => new ArrowFileReader(stream, memoryPool));
+        }
+
+        [Fact]
```
Comment (adamsitnik) [inline csharp/test/Apache.Arrow.Tests/ArrowFileReaderTests.cs]: nit: you could use theory here and add two test cases: true & false

File: csharp/test/Apache.Arrow.Tests/ArrowFileReaderTests.cs [hunk 11]
```diff
@@ -47,6 +48,55 @@ public void Ctor_LeaveOpenTrue_StreamValidOnDispose()
             Assert.Equal(0, stream.Position);
         }
 
+        [Fact]
+        public async Task Ctor_MemoryPool_AllocatesFromPool()
+        {
+            await Ctor_MemoryPool_AllocatesFromPoolHelper(
```
Comment (adamsitnik) [inline csharp/test/Apache.Arrow.Tests/ArrowFileReaderTests.cs]: nit: you can just return the task without awaiting it (the unit test runner should do it)

**Review by adamsitnik — COMMENTED** [review]
The performance improvements are really impressive! 🥇 

The code looks good to me, but I left some questions, mostly about the case when the memory owner is null.

I also wonder if we could somehow force the users to Dispose the Batch properly?

For the contributors to this project we could have sth like:

```cs
#if DEBUG
~RecordBatch()
{
   Debug.Fail("The Batch should have been Disposed in deterministic way using Dispose");
}
#endif
```

but for the end users maybe we should have some small package with static code analysis rule?

File: csharp/src/Apache.Arrow/ArrowBuffer.cs [hunk 12]
```diff
@@ -14,21 +14,37 @@
...
 {
-    public readonly partial struct ArrowBuffer: IEquatable<ArrowBuffer>
+    public readonly partial struct ArrowBuffer : IEquatable<ArrowBuffer>, IDisposable
     {
+        private readonly IMemoryOwner<byte> _memoryOwner;
+        private readonly ReadOnlyMemory<byte> _memory;
+
         public static ArrowBuffer Empty => new ArrowBuffer(Memory<byte>.Empty);
 
         internal ArrowBuffer(ReadOnlyMemory<byte> data)
```
Comment (pgovind) [inline csharp/src/Apache.Arrow/ArrowBuffer.cs]: Perhaps make this constructor public?

**Review by pgovind — COMMENTED** [review]
LGTM!

File: csharp/src/Apache.Arrow/ArrowBuffer.cs (hunk 12 above)
Comment (eerhardt) [reply csharp/src/Apache.Arrow/ArrowBuffer.cs]: I thought you were doing that in https://github.com/apache/arrow/pull/3971?  (BTW - I missed that you updated that PR, and approved it.)

File: csharp/src/Apache.Arrow/Memory/NativeMemoryAllocator.cs [hunk 13]
```diff
@@ -41,14 +42,10 @@ protected override Memory<byte> AllocateInternal(int length, out int bytesAlloca
 
             GC.AddMemoryPressure(bytesAllocated);
 
-            return manager.Memory;
-        }
+            // Ensure all allocated memory is zeroed.
```
[the remark this answers is not in the document]
Comment (chutchinson) [reply csharp/src/Apache.Arrow/Memory/NativeMemoryAllocator.cs]: > why do you zero the memory? I would expect that it might have some random values, but they are getting overwritten anyway by the reader later.

I was responsible for this. It was useful during development to identify padding when inspecting buffers (since all padding is zeroed), and at the time we desired the same memory allocation semantics as general .NET primitives. I still see the value in zeroing the memory, but perhaps it can be placed behind a flag and defaulted to false for a performance advantage (if any)?

**Comment by eerhardt** [comment]
> I also wonder if we could somehow force the users to Dispose the Batch properly?

It is `IDisposable`, I would assume any analyzer that enforced `IDisposable` should work. e.g. https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2213-disposable-fields-should-be-disposed

File: csharp/src/Apache.Arrow/Arrays/ArrayData.cs [hunk 14]
```diff
@@ -53,5 +54,24 @@ public sealed class ArrayData
...
+                {
+                    buffer.Dispose();
+                }
+            }
+
+            if (Children != null)
+            {
+                foreach (ArrayData child in Children)
+                {
+                    child?.Dispose();
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Arrays/ArrayData.cs]: I don't see any code that guards against having a `null` in the `Children` list, and it has a public constructor. So I assumed it can contain nulls. If we add a check, we can remove the `?.`.

File: csharp/src/Apache.Arrow/ArrowBuffer.Builder.cs [hunk 15]
```diff
@@ -109,20 +109,19 @@ public Builder<T> Clear()
...
-                }
+                int length = checked((int)BitUtility.RoundUpToMultipleOf64(_buffer.Length));
 
-                var memoryPool = pool ?? MemoryPool.Default.Value;
-                var memory = memoryPool.Allocate(length);
+                var memoryAllocator = allocator ?? MemoryAllocator.Default.Value;
+                var memoryOwner = memoryAllocator.Allocate(length);
 
-                Memory.CopyTo(memory);
+                if (memoryOwner != null)
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/ArrowBuffer.Builder.cs]: `memoryOwner` will be `null` when you pass in a `length == 0` to `memoryAllocator.Allocate(length)`.

Looking at this code more - it is using `_buffer.Length` which isn't correct. It should be using `Length`, which can be different (for example if I used the builder to create a `1000` elements, and then called `Clear()` and added `10` elements, I don't need a buffer big enough for `1000` elements.

File: csharp/src/Apache.Arrow/Arrays/Array.cs [hunk 16]
```diff
@@ -59,5 +59,19 @@ internal static void Accept<T>(T array, IArrowArrayVisitor visitor)
                     break;
             }
         }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Arrays/Array.cs]: `Array` has a `protected` constructor, which means anyone can derive from it.

Also - we are overriding the virtual Dispose in `ListArray.cs` to dispose its inner `Values` array.

---
In reply to: [281211354](https://github.com/apache/arrow/pull/3925#discussion_r281211354) [](ancestors = 281211354)

File: csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs (hunk 7 above)
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs]: What else would you expect to be returned when `0` gets passed into `Allocate`? Is allocating an object with an empty Memory worth it?

File: csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs (hunk 8 above)
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs]: The deserialization happens during `CreateArrowObjectFromMessage`. That method creates a `RecordBatch` and associated `IArrowArray` objects pointing into the memory here. The memory needs to be kept alive because how else would we remember all the values that we just read from the Stream?

File: csharp/src/Apache.Arrow/Memory/NativeMemoryAllocator.cs [hunk 17]
```diff
@@ -14,16 +14,17 @@
...
+    public class NativeMemoryAllocator : MemoryAllocator
     {
-        public NativeMemoryPool(int alignment = DefaultAlignment) 
+        public NativeMemoryAllocator(int alignment = DefaultAlignment) 
             : base(alignment) { }
 
-        protected override Memory<byte> AllocateInternal(int length, out int bytesAllocated)
+        protected override IMemoryOwner<byte> AllocateInternal(int length, out int bytesAllocated)
         {
             // TODO: Ensure memory is released if exception occurs.
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Memory/NativeMemoryAllocator.cs]: I think this TODO was referring to when an exception happens during this method (like if `GC.AddMemoryPressure` fails).

File: csharp/src/Apache.Arrow/Memory/NativeMemoryAllocator.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Memory/NativeMemoryAllocator.cs]: I was just moving this zeroing from the base class (which would happen for ANY allocator) to just this native memory case. This way a managed memory implementation doesn't need to re-zero the memory, if it doesn't want to. Can you open a JIRA for this?

File: csharp/src/Apache.Arrow/RecordBatch.cs (hunk 9 above)
Comment (eerhardt) [reply csharp/src/Apache.Arrow/RecordBatch.cs]: It isn't sealed, and it has a `public` constructor, so I felt it was best. If we seal the class, we can remove the full dispose pattern.

File: csharp/src/Apache.Arrow/Arrays/ArrayData.cs [hunk 18]
```diff
@@ -53,5 +54,24 @@ public sealed class ArrayData
...
+        {
+            if (Buffers != null)
+            {
+                foreach (ArrowBuffer buffer in Buffers)
+                {
+                    buffer.Dispose();
+                }
+            }
+
+            if (Children != null)
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Arrays/ArrayData.cs]: It's just used in these 2 places, is it really worth an extra method? Also, in one case it can have nulls and the other it can't (because `ArrowBuffer` is a `struct`).

File: csharp/src/Apache.Arrow/ArrowBuffer.Builder.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/ArrowBuffer.Builder.cs]: Fixing this is a little bigger than I thought it would be. So I separated it into a commit on its own.

File: csharp/test/Apache.Arrow.Tests/ArrowFileReaderTests.cs (hunk 10 above)
Comment (eerhardt) [reply csharp/test/Apache.Arrow.Tests/ArrowFileReaderTests.cs]: Good call. Fixed.

File: csharp/test/Apache.Arrow.Tests/ArrowFileReaderTests.cs (hunk 11 above)
Comment (eerhardt) [reply csharp/test/Apache.Arrow.Tests/ArrowFileReaderTests.cs]: With the above theory, I eliminated this helper method all together.

File: csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs [hunk 19]
```diff
@@ -62,7 +65,7 @@ protected override async ValueTask ReadSchemaAsync()
             await ValidateFileAsync().ConfigureAwait(false);
 
             int footerLength = 0;
-            await Buffers.RentReturnAsync(4, async (buffer) =>
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs]: If this is shown to be a problem, we can fix it in another change.

File: csharp/src/Apache.Arrow/ArrowBuffer.cs [hunk 20]
```diff
@@ -40,16 +56,21 @@ public ReadOnlySpan<byte> Span
...
-        public ArrowBuffer Clone(MemoryPool pool = default)
+        public ArrowBuffer Clone(MemoryAllocator allocator = default)
         {
             return new Builder<byte>(Span.Length)
                 .Append(Span)
-                .Build(pool);
+                .Build(allocator);
         }
 
         public bool Equals(ArrowBuffer other)
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/ArrowBuffer.cs]: I wonder if we should either remove `IEquatable` all together, or if we should delegate to the underlying `Memory`/`IMemoryOwner` for these instead. Comparing (and hashing) all the values in the buffer doesn't seem correct.

`Memory<T>` doesn't compare the values it points to, but instead checks if the underlying object/index/length are the same.

https://source.dot.net/#System.Private.CoreLib/shared/System/Memory.cs,500

[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/ArrowBuffer.cs]: @chutchinson - do you know why `ArrowBuffer` implements `IEquatable`? Is it a necessity?

File: csharp/src/Apache.Arrow/ArrowBuffer.Builder.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (adamsitnik) [reply csharp/src/Apache.Arrow/ArrowBuffer.Builder.cs]: Thank you for the explanation, it was not obvious to me.

Have you considered using the Null Object Pattern and returning an empty memory instead of null in such case?

File: csharp/src/Apache.Arrow/Memory/NativeMemoryAllocator.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (adamsitnik) [reply csharp/src/Apache.Arrow/Memory/NativeMemoryAllocator.cs]: Thanks for the explanation. If it shows up in the profiles in the next performance investigation we can consider removing it then. #resolved

File: csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs [hunk 21]
```diff
@@ -241,7 +244,7 @@ private async ValueTask ValidateMagicAsync()
 
             try
             {
-                await Buffers.RentReturnAsync(magicLength, async (buffer) =>
+                await ArrayPool<byte>.Shared.RentReturnAsync(magicLength, async (buffer) =>
```
[the remark this answers is not in the document]
Comment (adamsitnik) [reply csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs]: @chutchinson thanks for the explanation, cool name btw ;)

File: csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs (hunk 8 above)
Comment (adamsitnik) [reply csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs]: It makes sense to me now #solved

File: csharp/src/Apache.Arrow/Arrays/ArrayData.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (adamsitnik) [reply csharp/src/Apache.Arrow/Arrays/ArrayData.cs]: In 3, the nulls can be handled too with `?` (the generic implementation for struct should remove the null check) 

https://github.com/apache/arrow/blob/37a9b187276a5a68760becb816faeb0d44b00ac2/csharp/src/Apache.Arrow/RecordBatch.cs#L56-L58

File: csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs (hunk 7 above)
Comment (adamsitnik) [reply csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs]: I asked this question because I did not understand why it's returning a null, not because I thought it's wrong to do so.

**Review by adamsitnik — APPROVED** [review]
LGTM

File: csharp/src/Apache.Arrow/Memory/NativeMemoryAllocator.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Memory/NativeMemoryAllocator.cs]: I've opened https://issues.apache.org/jira/browse/ARROW-5276 to add this option.

File: csharp/src/Apache.Arrow/ArrowBuffer.Builder.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/ArrowBuffer.Builder.cs]: Thanks for the suggestion, @adamsitnik. I've opened https://issues.apache.org/jira/browse/ARROW-5277 to implement this.

File: csharp/src/Apache.Arrow/ArrowBuffer.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (eerhardt) [reply csharp/src/Apache.Arrow/ArrowBuffer.cs]: I've opened https://issues.apache.org/jira/browse/ARROW-5278 to address this.
